### PR TITLE
feat: add support for absolute URLs

### PR DESCRIPTION
**Pull request by thbil**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | NA
| License       | MIT

This PR enables absolute URLs validation.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # Change Log
 
+## 1.0.3
```
Comment (alanpoulain) [inline CHANGELOG.md:3]:
```suggestion
## 1.1.0
```

File: README.md [hunk 2]
```diff
@@ -132,7 +133,8 @@ class DocumentController extends AbstractController
         // Not passing the second argument will use the default expiration time (1 day by default).
         // return $this->urlSigner->sign($url);
 
-        // Will return a URL (more precisely a path) like this: /documents/42?expires=1611316656&signature=82f6958bd5c96fda58b7a55ade7f651fadb51e12171d58ed271e744bcc7c85c3
+        // Will return a path like this: /documents/42?expires=1611316656&signature=82f6958bd5c96fda58b7a55ade7f651fadb51e12171d58ed271e744bcc7c85c3
+        // Or a URL depending on what has been signed before
```
Comment (alanpoulain) [inline README.md:137]:
```suggestion
        // Or a URL depending on what has been signed before.
```

File: features/app/config/routes.yaml [hunk 3]
```diff
@@ -4,6 +4,10 @@ secured_document:
     defaults:
         _signed: true
 
-sign:
-    path: /sign
-    controller: CoopTilleuls\UrlSignerBundle\Tests\Controller\SignController
+sign-path:
+    path: /sign-path
+    controller: CoopTilleuls\UrlSignerBundle\Tests\Controller\SignPathController
```
Comment (alanpoulain) [inline features/app/config/routes.yaml:9]: Wouldn't it be better to have only one controller and a query parameter to control the generated signed URL?

File: tests/EventListener/ValidateSignedRouteListenerTest.php [hunk 4]
```diff
@@ -102,10 +107,24 @@ public function testValidateSignedRouteInvalidSignature(): void
...
+
         $this->signerProphecy->validate('/invalid-signature')->willReturn(false);
+        $this->signerProphecy->validate('http://test.org/invalid-signature')->willReturn(false);
 
         $this->dispatcher->dispatch($event);
 
         $this->signerProphecy->validate(Argument::any())->shouldHaveBeenCalledOnce();
     }
+
+    /** @return iterable<array-key, array<array-key, string>> */
```
Comment (alanpoulain) [inline tests/EventListener/ValidateSignedRouteListenerTest.php:119]:
```suggestion
    /** @return iterable<string, array<string, string>> */
```

File: tests/EventListener/ValidateSignedRouteListenerTest.php [hunk 5]
```diff
@@ -50,16 +50,21 @@ public function testSubscribedEvents(): void
...
+        $this->signerProphecy->validate($validUrl)->willReturn(true);
 
         $this->dispatcher->dispatch($event);
 
-        $this->signerProphecy->validate(Argument::any())->shouldHaveBeenCalledOnce();
+        if (0 === strpos($validUrl, '/')) {
+            $this->signerProphecy->validate(Argument::any())->shouldHaveBeenCalledOnce();
+        } else {
+            $this->signerProphecy->validate(Argument::any())->shouldHaveBeenCalledTimes(2);
+        }
```
Comment (alanpoulain) [inline tests/EventListener/ValidateSignedRouteListenerTest.php:67]:
```suggestion
        $isPath = 0 === strpos($validUrl, '/');
        $this->signerProphecy->validate(Argument::any())->shouldHaveBeenCalledTimes($isPath ? 1 : 2);
```

File: features/app/config/routes.yaml (hunk 3 above)
Comment (thbil) [reply features/app/config/routes.yaml:9]: I thought about it at first and it felt a little bit "hacky" to me, but I can change it if you want. You tell me!

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,11 @@
 # Change Log
 
+## 1.1.0
+
+### Bug Fixes
```
Comment (alanpoulain) [inline CHANGELOG.md:5]:
```suggestion
### New Features
```

File: features/app/src/Controller/SignController.php [hunk 7]
```diff
@@ -24,8 +26,11 @@ public function __construct(UrlSignerInterface $urlSigner)
         $this->urlSigner = $urlSigner;
     }
 
-    public function __invoke(): Response
+    public function __invoke(Request $request): Response
     {
-        return new Response($this->urlSigner->sign($this->generateUrl('secured_document', ['id' => 42]), 3));
+        /** @var int $isAbsoluteUrl */
+        $isAbsoluteUrl = $request->query->get('referenceType', 1);
```
Comment (alanpoulain) [inline features/app/src/Controller/SignController.php:32]: Why not using the retrieved type directly?

File: tests/Behat/UrlSignerContext.php [hunk 8]
```diff
@@ -39,6 +39,14 @@ public function iCreateASignedUrl(): void
         $this->client->request('GET', '/sign');
     }
 
+    /**
+     * @When I create an absolute signed URL
+     */
+    public function iCreateAnAbsoluteSignedUrl(): void
+    {
+        $this->client->request('GET', '/sign?referenceType=0');
```
Comment (alanpoulain) [inline tests/Behat/UrlSignerContext.php:47]: You can use the constant here.

File: features/app/src/Controller/SignController.php (hunk 7 above)
Comment (alanpoulain) [reply features/app/src/Controller/SignController.php:32]: Also the default value should be the constant.

**Review by alanpoulain — APPROVED** [review]

**Comment by alanpoulain** [comment]
Thank you @thbil.

**Pull request merged**